lua/plenary/async_lib/async.lua

```diff
@@ -0,0 +1,180 @@
+    local ret = {select(2, unpack(res))}
+
+    assert(stat, string.format("The coroutine failed with this message: %s", ret[1]))
+
+    if co.status(thread) == "dead" then
+      (callback or function() end)(unpack(ret))
+    else
+      assert(#ret == 1, "expected a single return value")
+      local returned_future = ret[1]
+      assert(type(returned_future) == "function", "type error :: expected func")
+      returned_future(step)
+    end
+  end
+
+  step()
+end
+
+-- use with CPS function, creates future factory
+-- must have argc for arity checking
+M.wrap = function(func, argc)
+  assert(type(func) == "function", "type error :: expected func, got " .. type(func))
+  assert(type(argc) == "number" or argc == "vararg", "expected argc to be a number or string literal 'vararg'")
+
+  return function(...)
+    local params = {...}
+
+    local function future(step)
+      if step then
+        if type(argc) == "number" then
+          params[argc] = step
+        else
+          table.insert(params, step) -- change once not optional
+        end
+        return func(unpack(params))
+      else
+        return co.yield(future)
+      end
+    end
+    return future
+  end
+end
+
+--- WIP
+local thread_loop_async = M.wrap(thread_loop, 2)
+
+-- many futures -> single future
+M.join = M.wrap(function(futures, step)
+  local len = #futures
+  local results = {}
+  local done = 0
+
+  if len == 0 then
+    return step(results)
+  end
+
+  for i, future in ipairs(futures) do
+    assert(type(future) == "function", "type error :: future must be function")
+    local callback = function(...)
+      results[i] = {...} -- should we set this to a table
+      done = done + 1
+      if done == len then
+        -- step(unpack(results))
+        step(results) -- should we unpack?
+      end
+    end
+    future(callback)
+  end
+end, 2)
+
+--- use this over running a future by calling it with no callback argument because it is more explicit
+M.run = function(future, callback)
+  future(callback or function() end)
+end
+
+M.run_all = function(futures, callback)
+  M.run(M.join(futures), callback)
+end
+
+M.await = function(future)
+  return future(nil)
+end
+
+M.await_all = function(futures)
+  assert(type(futures) == "table", "type error :: expected table")
+  return M.await(M.join(futures))
+end
+
+-- suspend co-routine, call function with its continuation (like call/cc)
+M.suspend = co.yield
+
+M.async = function(func)
+  return function(...)
+    local args = {...}
+    local function future(step)
+      if step == nil then
+        return func(unpack(args))
+      else
+        execute(future, step)
+      end
+    end
+    return future
+  end
+end
+
+-- converts an async function to callback based function
+M.convert = function(async_func)
+  return function(...)
+    local args = {...}
+    local callback = table.remove(args)
+    assert(type(callback) == "function" or type(callback) == "nil", "type error :: expected function as last vararg")
```

Comment:
this looks a bit brittle, `nil` cannot generally be passed as last arg. though is `M.convert` really needed for anything?

Reply:
the user might have not passed in anything so then it would be nil. `M.convert` just converts an async function to callback for using with old code that doesn't use async await yet.

Reply:
My lua knowledge might be lacking but if args is non-empty can `table.remove` ever return nil?

Reply:
if args is non-empty `table.remove` should not return `nil`

Reply:
This just confirms _exactly_ what I just said? Why would skipping callback be allowed just for function with no other args but not for any other function?

Reply:
Good catch, that needs to be fixed